_posts/2018-05-20-Y-The-Most-Beautiful-Idea-in-Computer-Science.md

```diff
@@ -0,0 +1,611 @@
+
+```js
+const refactoredFactorial = f => n => n === 0 ? 1 : n * f(n - 1);
+
+const Y = f => f(Y(f));
+
+const factorial = Y(refactoredFactorial);
+
+console.log(factorial(5));
+```
+
+Nope, just not yet. If you try to execute the snippet above, this will happen:
+
+```
+RangeError: Maximum call stack size exceeded
+    at Y (repl:1:11)
+    at Y (repl:1:18)
+    at Y (repl:1:18)
+    at Y (repl:1:18)
+    at Y (repl:1:18)
+    at Y (repl:1:18)
+    at Y (repl:1:18)
+    at Y (repl:1:18)
+    at Y (repl:1:18)
+    at Y (repl:1:18)
+```
+
+You will get a Stack Overflow because `Y` is calling itself infinitely. If you stop and replace `Y(f)` in the body of that function with its result, this will become obvious:
+
+```js
+const Y = f => f(Y(f));
+
+// This is what happens when we apply `f` to it:
+// Y(f) = f(Y(f));
+
+// That means we can replace every `Y(f)` with `f(Y(f))`
+// = f(Y(f));
+// = f(f(Y(f)));
+// = f(f(f(Y(f))));
+// And so on...
+```
+
+In fact, this very definition would work in lazy (a.k.a. [`non-strict`](https://en.wikipedia.org/wiki/Strict_programming_language)) languages, because we would not need to evaluate that `Y(f)` in the body of `Y` instantly, instead we would only evaluate it when needed. This means that instead of infinitely doing that "replacement" we've just done, we would get back the function returned by `f` with `Y(f)` being the first argument it takes (without having to evaluate it).
+
+Okay, so, in order to translate this to JavaScript we will need to delay the execution of `Y(f)` so that it does not get called immediately when we pass a function to `Y`.
+
+And what is the easiest way of delaying the execution of a function in `JavaScript`?
+
+A thunk! [A thunk is a function that wraps an expression to delay its evaluation](https://github.com/reduxjs/redux-thunk#whats-a-thunk).
+
+```js
+// calculation of 1 + 2 is immediate
+// x === 3
+let x = 1 + 2;
+
+// calculation of 1 + 2 is delayed
+// foo can be called later to perform the calculation
+// foo is a thunk!
+let foo = () => 1 + 2;
+```
+
+Does this remember you of anything you probably use on a daily-basis if you work with Redux? Exactly, [that's exacly what redux-thunk was made for](https://github.com/reduxjs/redux-thunk#motivation). In fact, [the very example above has been taken from their README](https://github.com/reduxjs/redux-thunk#whats-a-thunk).
+
+Now that we know what a thunk is, let's delay the execution of `f(Y(f))` by wrapping it into a function:
+
+```js
+const Y = f => x => f(Y(f))(x);
+```
+
+Since the function returned by `refactoredFactorial` will only take a single argument as a parameter, we can just wrap it inside a function that also takes a single argument and applies it to the result of `f(Y(f))`, which now will be a function that also takes a single argument instead of evaluating infinitely.
+
+Let's make this more obvious by imagining what happens when we invoke Y with `refactoredFactorial`:
+
+```js
+const refactoredFactorial = 
```

Comment:
`const refactoredFactorial = ?`

Reply:
This has been fixed in the full version. My fault for not pushing it all 😅